test/gtest/ucp/test_ucp_mem_type.cc

```diff
@@ -97,3 +97,70 @@ UCS_TEST_P(test_ucp_mem_type_alloc_before_init, xfer) {
 }
 
 UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_mem_type_alloc_before_init, all, "all")
+
+class test_ucp_cuda : public ucp_test {
+public:
+    static void get_test_variants(std::vector<ucp_test_variant>& variants) {
+        add_variant(variants, UCP_FEATURE_AM);
+    }
+};
+
+UCS_TEST_P(test_ucp_cuda, sparse_regions) {
+    const ucs_memory_type_t mem_type = UCS_MEMORY_TYPE_CUDA;
+    const size_t size = 4096;
+    const size_t count = 5;
+    ucs_status_t status;
+    void *ptr[count];
+
+    /* create contiguous CUDA registrations list */
+    for (int i = 0; i < count; i++) {
+        ptr[i] = mem_buffer::allocate(size, mem_type);
+
+        if ((i > 0) && (UCS_PTR_BYTE_OFFSET(ptr[i - 1], size) != ptr[i])) {
+            for (int j = 0; j < i; j++) {
+                mem_buffer::release(ptr[j], mem_type);
+            }
+            UCS_TEST_SKIP_R("failed to create contiguous CUDA registrations list");
+        }
+    }
+
+    /* make CUDA registrations list sparse*/
+    for (int i = 0; i < count; i++) {
+        if ((i & 1) == 0) {
+            mem_buffer::release(ptr[i], mem_type);
+        }
+    }
+
+    for (int i = 0; i < count; i++) {
+        if ((i & 1) == 1) {
+            ucp_mem_map_params_t params;
+            params.field_mask = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
+                                UCP_MEM_MAP_PARAM_FIELD_LENGTH;
+            params.address    = ptr[i];
+            params.length     = size;
+
+            ucp_mem_h memh;
```

Comment:
maybe map all regions first, to allow merges to happen, and only then pack the rkeys?